File: pinkmoe_index/src/hooks/shop/goods.ts
/*
 * @Author: coderzhaolu && izhaicy@163.com
 * @Date: 2022-07-21 14:16:37
 * @LastEditors: coderzhaolu && izhaicy@163.com
 * @LastEditTime: 2022-08-22 22:50:29
 * @FilePath: /pinkmoe_index/src/hooks/shop/goods.ts
 * @Description: https://github.com/Coder-ZhaoLu/pinkmoe   (如需用于商业用途或者二开，请联系作者捐助任意金额即可)
 * QQ:2419857357;支付宝:13135986153
 * Copyright (c) 2022 by coderzhaolu, All Rights Reserved.
 */
import { postView } from '/@/api/post'
import type { ResComment } from '/@/api/post/types'
import { useAppStore, useUserStore } from '/@/store'
import type { ResPost } from '/@/api/home/types'
import type { ResPage } from '/@/api/common/types'
import { getCommentList } from '/@/api/comment'
import type { ReqComment } from '/@/api/comment/types'
import { getGoodsItem } from '/@/api/goods'
import type { ReqGoodsItem, ResGoodsItem } from '/@/api/goods/types'

export const useGoodsItem = () => {
  const route = useRoute()
  const router = useRouter()
  const comment = ref()
  const cover = ref()
  const num = ref(0)
  const saleAmount = ref(0)
  const stock = ref(0)
  const goodsItem = ref<ResGoodsItem>()
  const commentList = ref<ResPage<Array<ResPost>>>()
  const user = useUserStore()
  const loading = ref<boolean>(false)
  const hasMore = ref<boolean>(true)
  const auth = useUserStore()
  const { siteBasic } = useAppStore()
  const { proxy } = getCurrentInstance()
  const formParams: ReqGoodsItem = reactive({
    goodsId: route.params.id,
  })

  const commentFormParams: ReqComment = reactive({
    postId: route.params.id,
    page: 1,
    pageSize: 12,
  })

  // 获取文章信息
  const getPost = async () => {
    goodsItem.value = await getGoodsItem(formParams)
    goodsItem.value.goods?.sizeRelation?.forEach((element) => {
      const v = JSON.parse(element.value)
      const vv = []
      v.forEach((ele, index) => {
        if (index === 0) {
          vv.push({
            value: ele,
            status: true,
          })
        }
        else {
          vv.push({
            value: ele,
            status: false,
          })
        }
      })
      element.value = vv
    })
    saleAmount.value = goodsItem.value.goods?.sizeValRelation?.[0].saleAmount
    stock.value = goodsItem.value.goods?.sizeValRelation?.[0].stock
    cover.value = goodsItem.value?.goods?.cover
  }

  const showComment = (commentMeta) => {
    commentFormParams.page = 1
    if (auth.isLogin) {
      comment.value.show(commentMeta)
    }
    else {
      proxy.$login({
        type: 'login',
        router,
        route,
      })
    }
  }

  const getComment = async () => {
    loading.value = true
    commentList.value = await getCommentList(commentFormParams)
    setTimeout(() => {
      loading.value = false
    }, 300)
  }

  const refreshComment = async () => {
    hasMore.value = true
    getComment()
  }

  const nextPage = async () => {
    loading.value = true
    commentFormParams.page! += 1
    const res = await getCommentList(commentFormParams)
    if (!res.list || res.list?.length <= 0)
      hasMore.value = false

    else
      commentList.value?.list?.push(...(res?.list as Array<ResComment>))

    setTimeout(() => {
      loading.value = false
    }, 300)
  }

  const share = (type) => {
    console.log(route)
    switch (type) {
      case 'qq':
        window.open(
          `https://sns.qzone.qq.com/cgi-bin/qzshare/cgi_qzshare_onekey?url=/${postItem.value?.post?.postId}&title=${postItem.value?.post?.title}&pics=${postItem.value?.post?.cover}&desc=&summary=&site=`,
          '_blank',
        )
        break
      case 'weibo':
        window.open(
          `https://service.weibo.com/share/share.php?url=/${postItem.value?.post?.postId}&title=${postItem.value?.post?.title}&pic=${postItem.value?.post?.cover}&searchPic=true`,
          '_blank',
        )
        break
      case 'weixin':
        window.open(
          `https://service.weibo.com/share/share.php?url=/${postItem.value?.post?.postId}&title=${postItem.value?.post?.title}&pic=${postItem.value?.post?.cover}&searchPic=true`,
          '_blank',
        )
        break
      case 'bold':
        window.open(
          `http://tieba.baidu.com/f/commit/share/openShareApi?url=/${postItem.value?.post?.postId}&title=${postItem.value?.post?.title}&desc=&comment=&pic=${postItem.value?.post?.cover}&red_tag=j2697365826`,
          '_blank',
        )
        break
      default:
        window.open(
          `https://service.weibo.com/share/share.php?url=/${postItem.value?.post?.postId}&title=${postItem.value?.post?.title}&pic=${postItem.value?.post?.cover}&searchPic=true`,
          '_blank',
        )
        break
    }
  }

  const changeSizeVal = (index, size) => {
    size.forEach((element) => {
      element.status = false
    })
    size[index].status = true

    // 处理商品规格变化
    const sizeValPool = []
    goodsItem.value?.goods?.sizeRelation?.forEach((e) => {
      e.value.forEach((ee) => {
        if (ee.status === true)
          sizeValPool.push(ee.value)
      })
    })

    goodsItem.value?.goods?.sizeValRelation?.forEach((e) => {
      if (e.value == JSON.stringify(sizeValPool)) {
        stock.value = e.stock
        saleAmount.value = e.saleAmount
      }
    })
  }

  onMounted(() => {
    postView({ postId: route.params.id })
    getPost()
    getComment()
  })

  return {
    goodsItem,
    user,
    route,
    siteBasic,
    comment,
    commentList,
    hasMore,
    loading,
    stock,
    saleAmount,
    cover,
    num,
    share,
    nextPage,
    changeSizeVal,
    getComment,
    showComment,
    refreshComment,
  }
}

File: pinkmoe_server/model/request/goods.go
/*
 * @Author: coderzhaolu && izhaicy@163.com
 * @Date: 2022-06-22 11:13:07
 * @LastEditors: coderzhaolu && izhaicy@163.com
 * @LastEditTime: 2022-08-07 08:53:45
 * @FilePath: /pinkmoe_server/model/request/post.go
 * @Description: https://github.com/Coder-ZhaoLu/pinkmoe 
 * 问题反馈qq群:749150798
 * xanaduCms程序上所有内容(包括但不限于 文字，图片，代码等)均为指针科技原创所有，采用请注意许可
 * 请遵循 “非商业用途” 协议。商业网站或未授权媒体不得复制内容，如需用于商业用途或者二开，请联系作者捐助任意金额即可，我们将保存所有权利。
 * Copyright (c) 2022 by 指针科技, All Rights Reserved.
 */
package request

import (
	"server/global"
	"server/model"

	uuid "github.com/satori/go.uuid"
)

// SearchGoodsParams api分页条件查询及排序结构体
type SearchGoodsParams struct {
	PageInfo
	GoodsId       string `json:"goodsId" form:"goodsId" gorm:"not null;unique;primary_key;comment:商品ID"` // 商品ID
	Title         string `json:"title" form:"title" gorm:"comment:商品标题"`                                 // 商品标题
	Content       string `json:"content" form:"content" gorm:"type:text;comment:商品内容"`                   // 商品内容
	Category      string `json:"category" form:"category" gorm:"comment:商品分类"`                           // 商品分类
	Author        string `json:"author" form:"author" gorm:"comment:商品作者"`                               // 商品作者
	Cover         string `json:"cover" form:"cover" gorm:"comment:商品封面"`                                 // 商品封面
	Type          string `json:"type" form:"type" gorm:"comment:商品类型"`                                   // 商品类型
	View          int    `json:"view" form:"view" gorm:"comment:商品查看"`                                   // 商品查看
	CommentStatus string `json:"commentStatus" form:"commentStatus" gorm:"default:true;comment:评论状态"`    // 评论状态
	Status        string `json:"status" form:"status" gorm:"comment:商品状态"`                               // 商品状态
	OrderKey      string `json:"orderKey" form:"orderKey" gorm:"default:updated_at;"`                    // 排序
	Desc          bool   `json:"desc" form:"desc"`                                                       // 排序方式:升序false(默认)|降序true
}

type ReqGoods struct {
	CreatedAt     global.XdTime  // 创建时间
	UpdatedAt     global.XdTime  // 更新时间
	DeletedAt     *global.XdTime `gorm:"index" json:"-"`                                                         // 删除时间
	GoodsId       string         `json:"goodsId" form:"goodsId" gorm:"not null;unique;primary_key;comment:商品ID"` // 商品ID
	Title         string         `json:"title" form:"title" gorm:"comment:商品标题"`                                 // 商品标题
	Content       string         `json:"content" form:"content" gorm:"type:text;comment:商品内容"`                   // 商品内容
	Category      string         `json:"category" form:"category" gorm:"comment:商品分类"`                           // 商品分类
	Author        uuid.UUID      `json:"author" form:"author" gorm:"comment:商品作者"`                               // 商品作者
	Cover         string         `json:"cover" form:"cover" gorm:"comment:商品封面"`                                 // 商品封面
	Type          string         `json:"type" form:"type" gorm:"comment:商品类型"`                                   // 商品类型
	View          int            `json:"view" form:"view" gorm:"comment:商品查看"`                                   // 商品查看
	CommentStatus string         `json:"commentStatus" form:"commentStatus" gorm:"default:true;comment:评论状态"`    // 评论状态
	Status        string         `json:"status" form:"status" gorm:"comment:商品状态"`                               // 商品状态
}

type XdGoodsSize struct {
	CreatedAt global.XdTime  // 创建时间
	UpdatedAt global.XdTime  // 更新时间
	DeletedAt *global.XdTime `gorm:"index" json:"-"`                                                         // 删除时间
	Key       string         `json:"key" form:"key" gorm:"comment:规格名称"`                                     // 规格名称
	Value     []string       `json:"value" form:"value" gorm:"comment:规格配置"`                                 // 规格配置
	GoodsId   string         `json:"goodsId" form:"goodsId" gorm:"not null;unique;primary_key;comment:商品ID"` // 商品ID
}

type XdGoodsSizeVal struct {
	CreatedAt  global.XdTime  // 创建时间
	UpdatedAt  global.XdTime  // 更新时间
	DeletedAt  *global.XdTime `gorm:"index" json:"-"`                                                         // 删除时间
	Stock      int            `json:"stock" form:"stock" gorm:"comment:商品库存"`                                 // 商品库存
	Value      []string       `json:"value" form:"value" gorm:"comment:商品值"`                                  // 商品值
	SaleAmount int            `json:"saleAmount" form:"saleAmount" gorm:"comment:商品零售价"`                      // 商品零售价
	GoodsId    string         `json:"goodsId" form:"goodsId" gorm:"not null;unique;primary_key;comment:商品ID"` // 商品ID
}

// CreateGoodsParams api分页条件查询及排序结构体
type CreateGoodsParams struct {
	GoodsId         string               `json:"goodsId" form:"goodsId" gorm:"not null;unique;primary_key;comment:商品ID"` // 商品ID
	Title           string               `json:"title" form:"title" gorm:"comment:商品标题"`                                 // 商品标题
	Content         string               `json:"content" form:"content" gorm:"type:text;comment:商品内容"`                   // 商品内容
	Category        string               `json:"category" form:"category" gorm:"comment:商品分类"`                           // 商品分类
	Author          uuid.UUID            `json:"author" form:"author" gorm:"comment:商品作者"`                               // 商品作者
	Cover           string               `json:"cover" form:"cover" gorm:"comment:商品封面"`                                 // 商品封面
	Type            string               `json:"type" form:"type" gorm:"comment:商品类型"`                                   // 商品类型
	View            int                  `json:"view" form:"view" gorm:"comment:商品查看"`                                   // 商品查看
	CommentStatus   string               `json:"commentStatus" form:"commentStatus" gorm:"default:true;comment:评论状态"`    // 评论状态
	Status          string               `json:"status" form:"status" gorm:"comment:商品状态"`                               // 商品状态
	SizeRelation    []XdGoodsSize        `json:"sizeRelation" gorm:"foreignKey:GoodsId;references:GoodsId;"`
	SizeValRelation []XdGoodsSizeVal     `json:"sizeValRelation" gorm:"foreignKey:GoodsId;references:GoodsId;"`
	GoodsImg        []model.XdUploadFile `json:"goodsImg" form:"goodsImg"`
}
